apis/Google.Cloud.AspNetCore.DataProtection.SecretManager/Google.Cloud.AspNetCore.DataProtection.SecretManager/CloudSecretManagerXmlRepository.cs

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at 
+//
+// https://www.apache.org/licenses/LICENSE-2.0 
+//
+// Unless required by applicable law or agreed to in writing, software 
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Api.Gax.ResourceNames;
+using Google.Cloud.SecretManager.V1;
+using Google.Protobuf;
+using Grpc.Core;
+using Microsoft.AspNetCore.DataProtection.Repositories;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Xml.Linq;
+using Secret = Google.Cloud.SecretManager.V1.Secret;
+
+namespace Google.Cloud.AspNetCore.DataProtection.SecretManager;
+
+/// <summary>
+/// Implementation of <see cref="IXmlRepository"/> that stores the protected elements in a Google Cloud Secret Manager.
+/// This class is configured by <see cref="GoogleCloudDataProtectionBuilderExtensions.PersistKeysToGoogleCloudSecretManager(Microsoft.AspNetCore.DataProtection.IDataProtectionBuilder, string, string)"/>
+/// (and other overloads).
+/// </summary>
+internal class CloudSecretManagerXmlRepository : IXmlRepository
+{
+    private readonly SecretManagerServiceClient _secretManagerServiceClient;
+    private readonly SecretName _secretName;
+    private readonly ProjectName _projectName;
+    private readonly SecretVersionName _version;
+
+    internal CloudSecretManagerXmlRepository(SecretManagerServiceClient client, string secretName, string projectId)
+    {
+        _secretManagerServiceClient = GaxPreconditions.CheckNotNull(client, nameof(client));
+        GaxPreconditions.CheckNotNull(secretName, nameof(secretName));
+        GaxPreconditions.CheckNotNull(projectId, nameof(projectId));
+        _projectName = new ProjectName(projectId);
+        _secretName = new SecretName(projectId, secretName);
+
+        // Version is populated in such a way that it always refers to the most latest version.
+        // For more information on version naming refer to
+        // https://cloud.google.com/dotnet/docs/reference/Google.Cloud.SecretManager.V1/latest/Google.Cloud.SecretManager.V1.AccessSecretVersionRequest#Google_Cloud_SecretManager_V1_AccessSecretVersionRequest_Name
+        _version = SecretVersionName.FromProjectSecretSecretVersion(projectId, secretName, "latest");
+    }
+
+    /// <inheritdoc />
+    public IReadOnlyCollection<XElement> GetAllElements()
+    {
+        if (!SecretExists())
```

Comment:
Rather than make an RPC just to check if it exists, can we call AccessSecretVersion unconditionally, and catch whatever failure we get if it doesn't exist?

Reply:
done